File: ForexTester/ForexTester/Strategies/S1a.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace ForexTester.Strategies
{
    class S1A : BaseStrategy
    {
        public S1A(List<Candle> candles)
            : base(candles)
        {
        }

        public void ComputeStrategy(object selectedItem)
        {
            string filename = selectedItem + "." + ResultFileExtension;
            var bestReturn = -120.0;
            var candlesCount = Candles.Count;
            var bestCalmar = 0.0;
            var bestMa = 0;
            using (var writer = new StreamWriter(filename, false))
            {
                writer.WriteLine("results");
            }
            //paramALength: 9 paramAVolLength: 26 paramABuffer: -0.00416 paramAVolThreshold: 100 paramADuration: 16 paramASL: 0.00576 Calmar: 2.6312
            //const int paramALength=24;
            for (var paramALength = 10; paramALength <= 1000; paramALength += 1)//liczba swiec wstecz do obliczenia maksimum
            {
                var means = new List<double>(candlesCount) { Candles.First().Close };
                var kon = candlesCount - 1;
                for (var i = 1; i < kon; i++)
                    means.Add(GetCandlesRange(i, paramALength).Average(candle => candle.Close));
                //CreateStatus("paramALength: {0}", paramALength);
                //AppendStatus("  paramAVolLength: {0}\r\n", paramAVolLength);
                var sumRa = new List<double>(candlesCount);
                var returnValues = new List<double>(candlesCount);
                var drawdowns = new List<double>(candlesCount);
                var candlesBegin = paramALength + 1;
                var positionCount = 0; //liczba otwieranych pozycji
                //var stopLossCount = 0; //liczba stop lossów
                var lastCandle = kon;
                for (int i = 0; i < candlesBegin; i++)
                {
                    sumRa.Add(0);
                    returnValues.Add(0);
                }
                for (int i = candlesBegin; i < lastCandle; i++)
                {
                    returnValues.Add(0);
                    if (means[i] < Candles[i].Close)
                    {
                        returnValues[i] = Candles[i + 1].Close - Candles[i + 1].Open;
                        //zysk z i-tej pozycji long zamykanej na zamknięciu po 1 kroku
                        positionCount++;
                    }
                    sumRa.Add(sumRa[i - 1] + returnValues[i]);  //krzywa narastania kapitału
                }


                var recordReturn = 0.0;  //rekord zysku
                var recordDrawdown = 0.0;  //rekord obsuniecia
                for (int j = 0; j < lastCandle; j++)
                {
                    if (sumRa[j] > recordReturn)
                    {
                        recordReturn = sumRa[j];
                    }
                    drawdowns.Add(sumRa[j] - recordReturn);//róznica pomiedzy bieżącą wartoscia kapitału skumulowanego a dotychczasowym rekordem
                    if (drawdowns[j] < recordDrawdown)
                        recordDrawdown = drawdowns[j];  //obsuniecie maksymalne
                }

                //wyniki końcowe
                var sumReturn = sumRa[lastCandle - 1];
                var calmar = -sumReturn / recordDrawdown;  //wskaznik Calmara
                if (bestReturn < sumReturn)
                {
                    bestReturn = sumReturn; bestCalmar = calmar;
                    bestMa = paramALength;
                    using (var writer = new StreamWriter(filename, true))
                    {
                        writer.WriteLine("{0}\t{1}\t{2}",
                                GetDouble(sumReturn), GetDouble(paramALength),
                                GetDouble(calmar));
                    }
                }
            }
            using (var writer = new StreamWriter(filename, true))
            {
                writer.WriteLine("{0}\t{1}\t{2}",
                        GetDouble(bestReturn), GetDouble(bestMa),
                        GetDouble(bestCalmar));
            }

        }
    }
}

File: Article/book/matlabs/S1a.m
function [] = S1a (C,ma,rynek)
%(C) Antoni Wiliñski 2013
%skrypt strategii S1 w projekcie TEWI
%S1a - pierwszy kwadrant - oznaczajacy za³o¿enie, ze trend jest rosnacy i
%nale¿y otwierac pozycje d³ugie wg zasady Buy Stop

candlesCount = size(C,1);

% candlesCount(1)=5000;
bestSumRa = -Inf;
bestCalmara = -Inf;

for ma=ma
    means=zeros(1, candlesCount);
    kon=candlesCount-1;
    for i=2:kon
        means(i) = mean(C(i-min(i-1,ma):i,4));
    end
    sumRa=zeros(1,6000);
    Ra=zeros(1,6000);
    pocz=ma+1;
    if size(C,1)<5000
        kon = 2047;
    else
        kon=5000-1;
    end
    la=0; %liczba otwieranych pozycji

    for i=pocz:kon
        if means(i)<C(i,4)
       % if mean(C(i-ma:i,4))<C(i,4) && mean(C(i-ma:i,4))>C(i-1,4)
            Ra(i)=C(i+1,4)-C(i+1,1); %zysk z i-tej pozycji long zamykanej po jednym kroku (na koñcu swiecy, w której nastapi³o otwarcie)
            la=la+1;
        end
        sumRa(i)=sum(Ra(pocz:i));  %krzywa narastania kapita³u
    end
    
    %obliczenie wspó³czynnika Calmara (stosunku zysku koñcowego do najwiekszego
    %obsuniecia)
    
    recZ=0;  %rekord zysku
    recO=0;  %rekord obsuniecia
    
    for j=1:kon
        if sumRa(j)>recZ
            recZ=sumRa(j);
        end
        dZ(j)=sumRa(j)-recZ; %róznica pomiedzy bie¿¹c¹ wartoscia kapita³u skumulowanego a dotychczasowym rekordem
        if dZ(j)<recO
            recO=dZ(j);  %obsuniecie maksymalne
        end 
    end
    %wyniki koñcowe
    Calmar=-sumRa(kon)/recO;  %wskaznik Calmara
    if sumRa(kon)>bestSumRa
        bestSumRa = sumRa(kon);
        bestSumAll = sumRa;
        bestCalmara = Calmar;
        bestLa = la;  %liczba otwartych pozycji
        bestLap=la/(kon-pocz);
        bestMa = ma;
        disp(['Zysk: ' num2str(sumRa(kon)) ' dla ma: ' num2str(ma)]);
    end              
end


ToTable = [round2(bestSumRa,0.001) round2(bestCalmara,0.001) bestMa bestLa];
save(['texRa_' rynek '.txt'],'ToTable','-ascii','-double');

hFig = figure;
set(hFig, 'Position', [200 200 840 400]);
bestSumAll = bestSumAll(bestSumAll~=0);
plot(bestSumAll);
xlim([0 length(bestSumAll)]);
xlabel('Swiece');
ylabel('Skumulowany zysk');
set(hFig, 'PaperPositionMode','auto') ;  %# WYSIWYG
print(hFig,'-depsc', '-r0',[mfilename '_' rynek]);
